PerfettoCds/Pipeline/Tables/PerfettoGenericEventTable.cs

```diff
@@ -0,0 +1,218 @@
+
+        private static readonly ColumnConfiguration TypeColumn = new ColumnConfiguration(
+            new ColumnMetadata(new Guid("{01d2b15f-b0fc-4444-a240-0a96f62c2c50}"), "Type", "Type of the event"),
+            new UIHints { Width = 70 });
+
+        public static void BuildTable(ITableBuilder tableBuilder, IDataExtensionRetrieval tableData)
+        {
+            // We dynamically adjust the column headers
+            // This is the max number of fields we can expect for this table
+            int maxFieldCount = Math.Min(AbsoluteMaxFields, tableData.QueryOutput<int>(
+                new DataOutputPath(PerfettoPluginConstants.GenericEventCookerPath, nameof(PerfettoGenericEventCooker.MaximumEventFieldCount))));
+
+            // Get data from the cooker
+            var events = tableData.QueryOutput<ProcessedEventData<PerfettoGenericEvent>>(
+                new DataOutputPath(PerfettoPluginConstants.GenericEventCookerPath, nameof(PerfettoGenericEventCooker.GenericEvents)));
+
+            // Start construction of the column order. Pivot on process and thread
+            List<ColumnConfiguration> allColumns = new List<ColumnConfiguration>() 
+            {
+                ProcessNameColumn,
+                ThreadNameColumn,
+                TableConfiguration.PivotColumn, // Columns before this get pivotted on
+                EventNameColumn,
+                DurationColumn,
+                CategoryColumn,
+                TypeColumn 
+            };
+
+            var tableGenerator = tableBuilder.SetRowCount((int)events.Count);
+            var genericEventProjection = new GenericEventProjection(events);
+
+            // Add all the data projections
+            var processNameColumn = new BaseDataColumn<string>(
+                ProcessNameColumn,
+                genericEventProjection.Compose((genericEvent) => genericEvent.Process));
+            tableGenerator.AddColumn(processNameColumn);
+
+            var threadNameColumn = new BaseDataColumn<string>(
+                ThreadNameColumn,
+                genericEventProjection.Compose((genericEvent) => genericEvent.Thread));
+            tableGenerator.AddColumn(threadNameColumn);
+
+            var eventNameColumn = new BaseDataColumn<string>(
+                EventNameColumn,
+                genericEventProjection.Compose((genericEvent) => genericEvent.EventName));
+            tableGenerator.AddColumn(eventNameColumn);
+
+            var timestampColumn = new BaseDataColumn<Timestamp>(
+                TimestampColumn,
+                genericEventProjection.Compose((genericEvent) => genericEvent.Timestamp));
+            tableGenerator.AddColumn(timestampColumn);
+
+            var durationColumn = new BaseDataColumn<long>(
+                DurationColumn,
+                genericEventProjection.Compose((genericEvent) => genericEvent.Duration));
+            tableGenerator.AddColumn(durationColumn);
+
+            var categoryColumn = new BaseDataColumn<string>(
```

Comment:
I wonder how we can separate these categories out such that WPA could group/aggregate on them?

Reply:
Not sure I follow. Doesn't WPA group on them if you make it a pivot column?

Reply:
We should brainstorm with @trgibeau on how to represent this when pivoting. Today if you make this a pivot column this is just pivoted as a string. Categories with only 1 item will be grouped together. Categories with multiple items (CSV separated) will only be grouped where all the categories are the same.

We could sort categories say alphabetically and break into Category1, Category2 although obviously not a perfect solution.

Ideally if we could split on CSV and represent this as an array and feed it to the UI to "smart" pivot that would be ideal. Basically want to be able to pivot on a single category at a time even with this more complex multi-category system.

Example of categories from Chromium:
content
toplevel.flow
input,benchmark
input,benchmark,devtools.timeline
gpu,toplevel.flow
renderer,benchmark,rail

Reply:
Will keep thinking about how these can be grouped. Need to learn more about WPA table capabilities.

Reply:
What about a hierarchical column similar to stacks? 
Allows for multiple levels, depends on how the categories are defined in the first place